netty/src/main/java/io/grpc/netty/AbstractNettyHandler.java

```diff
@@ -114,69 +115,91 @@ private void sendInitialConnectionWindow() throws Http2Exception {
+      if (!autoTuneFlowControlOn) {
+        return;
+      }
+      pingReturn++;
+      Http2LocalFlowController fc = decoder().flowController();
+      // Calculate new window size by doubling the observed BDP, but cap at max window
+      int targetWindow = Math.min(getDataSincePing() * 2, MAX_WINDOW_SIZE);
+      setPinging(false);
+      int currentWindow = fc.initialWindowSize(connection().connectionStream());
+      if (targetWindow > currentWindow) {
+        int increase = targetWindow - currentWindow;
+        fc.incrementWindowSize(connection().connectionStream(), increase);
+        fc.initialWindowSize(targetWindow);
+        Http2Settings settings = new Http2Settings();
+        settings.initialWindowSize(targetWindow);
+        frameWriter().writeSettings(ctx(), settings, ctx().newPromise());
+      }
+
+    }
+
+    private boolean isPinging() {
       return pinging;
     }
 
-    public void setPinging(boolean pingOut) {
+    private void setPinging(boolean pingOut) {
       pinging = pingOut;
     }
 
-    public void sendPing(ChannelHandlerContext ctx) {
+    private void sendPing(ChannelHandlerContext ctx) {
       setDataSizeSincePing(0);
       if (payloadBuff == null) {
         initializePayloadBuf();
       }
       payloadBuff.retain();
-      encoder().writePing(ctx, false, payloadBuff, ctx.newPromise());
+      encoder().writePing(ctx, false, payloadBuff.slice(), ctx.newPromise());
       pingCount++;
     }
 
-    public void updateWindow() throws Http2Exception {
-      pingReturn++;
-      // only perform window updates if flowControlOn. Otherwise, pings are essentially ignored
-      if (flowControlOn) {
-        Http2LocalFlowController fc = decoder().flowController();
-        // Calculate new window size by doubling the observed BDP, but cap at max window
-        int targetWindow = Math.min(getDataSincePing() * 2, MAX_WINDOW_SIZE);
-        setPinging(false);
-        int currentWindow = fc.initialWindowSize(connection().connectionStream());
-        if (targetWindow > currentWindow) {
-          int increase = targetWindow - currentWindow;
-          fc.incrementWindowSize(connection().connectionStream(), increase);
-          fc.initialWindowSize(targetWindow);
-          Http2Settings settings = new Http2Settings();
-          settings.initialWindowSize(targetWindow);
-          frameWriter().writeSettings(ctx(), settings, ctx().newPromise());
-        }
-      }
-    }
-
-    public void incrementDataSincePing(int increase) {
+    private void incrementDataSincePing(int increase) {
       int currentSize = getDataSincePing();
       setDataSizeSincePing(currentSize + increase);
     }
 
-    public int payload() {
-      return BDP_MEASUREMENT_PING;
-    }
-
-    public int maxWindow() {
-      return MAX_WINDOW_SIZE;
+    private void initializePayloadBuf() {
+      payloadBuff =
+          unreleasableBuffer(ctx().alloc().directBuffer(8).writeLong(BDP_MEASUREMENT_PING))
```

Comment:
This is still leaking. `payloadBuff` should be static, so you only ever create one instance. You can also initialize it statically (so no need for `initializePayloadBuf()`) and make the variable `final`